### ISSUE TITLE
[P1] Question on arithmetic reasoning results

### ISSUE THREAD

**Issue by clarenceluo78** [stanfordnlp/pyreft#42]
Hi there, thank you for sharing this interesting work!

I have one question regarding the arithmetic reasoning experiment: the baseline results are taken from Hu et al. [2023], where it is noted that in their setting the number of epochs is set to 3 on the math_10k, while in the loreft setting, the epoch is set to 12. I wonder whether this constitutes a fair comparison with the baseline. Do you have any insights on this? Thanks!

**Comment by frankaging** [comment]
hey @clarenceluo78 thanks for your comments! This is a really good point to raise!

Yes, we use different hyperparameters compared to previous PEFTs. The key difference is that ReFT requires a more comprehensive hyperparameter search process, as it involves a broader range of hyperparameters (e.g., the representation layers where interventions are applied), which leads to a reduction in the number of trainable parameters.

Regarding a fair comparison: Given the significant differences between ReFT and PEFTs, we argue that our approach is fair. Especially during hyperparameter tuning, we ensure not to use the real test sets (e.g., from LLM-Adaptor) for searching. This contrasts with the practice described in the LLM-Adaptor paper (refer to page 5, section 4.1, or Figure 3), where they selected the best configuration based on performance in a math reasoning task on a 7B model. We believe that, generally, this approach (i.e., searching hyperparameters, even just the learning rate, or batch size) is not a good practice for PEFTs (we call this out in our section 7).

In our view, a better practice is to select a separate dataset (possibly from a slightly different domain), conduct **extensive hyperparameter tuning**, and then apply the optimized set of hyperparameters to the actual benchmarks **without further adjustments**. This is the methodology we adopt, tuning our hyperparameters using the last 300 training examples from GSM8K. The decision to run for 12 epochs, among other settings, is derived from our tuning experiments.

A side note: Given that LoRA introduces a new variant almost every week or month, the commonly used hyperparameters (e.g., rank=8, epoch=3, etc.) seem to be already optimized for LoRA or its variants. However, we do not believe this applies to ReFT.

**Comment by frankaging** [comment]
Hey @clarenceluo78 **We rerun our main experiments with 3 seeds on DoRA/LoRA 3 epoch + batch size 16** settings just to show additional insights. Please be aware that **we do still think running with different epoch numbers and batch size are okay** as long as there is **no "hyperparameter tuning over test set" thing**! Thanks again for raising the issue. 

Results are attached if you are interested in them! We will include these results in our mega update next week along with other stuff.

<img width="964" alt="Screenshot 2024-04-12 at 9 51 07 AM" src="https://github.com/stanfordnlp/pyreft/assets/15223704/48e7468f-eab7-49d7-a183-4c678d043814">

<img width="960" alt="Screenshot 2024-04-12 at 9 50 41 AM" src="https://github.com/stanfordnlp/pyreft/assets/15223704/1f12dfdb-3c36-47ae-8555-22640613e6a5">
